sql: fix sql activity cache for large cardinality

maryliag opened the pull request:
Note to reviewers: first commit being reviewed on https://github.com/cockroachdb/cockroach/pull/112609

---
Recreated from #112071

tldr: fix activity update job failing from conflict, caching to much info when their is high cardinatlity in the stats, and fix cache to store the correct top stats.

1. Fix activity update failure from upserting same row multiple times. This was caused by 2 nodes having the same statement, but one only executed it 2 or 3 time and the other node executed it enough to generate an index recommendation. The different index recommendation caused duplicate rows based on primary key to be inserted. It now uses index recommendation from the merged stats which uses the latest existing
recommendation (we don't update if there is no recommendation).
2. Customers can have transactions that have over 1k unique statement fingerprint ids. This makes it unreasonable to cache all of the statement information for all the top transactions. This commit removes the logic to cache all the statement information for the top transactions which could cause the table to grow to over 100k for a single hour. This requires the call to go the statistics table to get all the necessary statement information for the txn details.
3. The activity cache logic was not correctly selecting the top CPU json column. It was using the computed index column name instead of the json field name. The values shown on the UI were correct because the endpoint still aggregated several columns using the aggregated stats rather than the separate column.
4. The activity cache logic was using the execution count instead of the statistics count to select the top queries. In most cases this still generates the correct top queries, and the UI uses the aggregated stats which showed the correct value.

Fixes: #111780
Fixes: #111224

Release note (sql change): Fix the SQL activity update job to avoid conflicts on update, reduces the amount of data cached to just what the overview page requires, and fix the correctess of the top queries.

maryliag committed:
sql: fix sql activity cache for large cardinality

Recreated from #112071

tldr: fix activity update job failing from conflict, caching to much
info when their is high cardinatlity in the stats, and fix cache to
store the correct top stats.

1. Fix activity update failure from upserting same row multiple times.
This was caused by 2 nodes having the same statement, but one only
executed it 2 or 3 time and the other node executed it enough to
generate an index recommendation. The different index recommendation
caused duplicate rows based on primary key to be inserted. It now uses
index recommendation from the merged stats which uses the latest existing
recommendation (we don't update if there is no recommendation).
2. Customers can have transactions that have over 1k unique statement
fingerprint ids. This makes it unreasonable to cache all of the
statement information for all the top transactions. This commit removes
the logic to cache all the statement information for the top
transactions which could cause the table to grow to over 100k for a
single hour. This requires the call to go the statistics table to get
all the necessary statement information for the txn details.
3. The activity cache logic was not correctly selecting the top CPU json
column. It was using the computed index column name instead of the json
field name. The values shown on the UI were correct because the endpoint
still aggregated several columns using the aggregated stats rather than
the separate column.
4. The activity cache logic was using the execution count instead of the
statistics count to select the top queries. In most cases this still
generates the correct top queries, and the UI uses the aggregated stats
which showed the correct value.

Fixes: #111780
Fixes: #111224

Release note (sql change): Fix the SQL activity update job to avoid
conflicts on update, reduces the amount of data cached to just what the
overview page requires, and fix the correctess of the top queries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112350)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @maryliag)

___
*[`-- commits` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Ngf8th60zvxbJ91tQJ6):*
This fixes #111224, right?
___
*[`pkg/server/combined_statement_stats.go` line 1016 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Ngf8b0b9pbhUoCnUqLi:-Ngf8b0b9pbhUoCnUqLj:b9lza7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/server/combined_statement_stats.go#L1016)):*
> ```Go
> // This does not use the activity tables because the statement information is
> // aggregated to remove the transaction fingerprint id to keep the size of the
> // statement_activity manageable when transaction have over 1k+ statement ids.
> ```

nit: "transaction have".
___
*[`pkg/sql/sql_activity_update_job.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfEojI6wALYxKzhqDV:-NgfEojI6wALYxKzhqDW:be28d03) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L368)):*
> ```Go
> 		// each of execution_count, total execution time, service_latency,cpu_sql_nanos,
> 		// contention_time and insert into transaction_activity table.
> 		// Up to 3000 rows (sql.stats.activity.top.max * 5) may be added to
> ```

Why did this change? Because we only need 5 columns? Also, do we need to reduce `numberOfTopColumns` to `5` too? Also, `s/3000/2500/`.
___
*[`pkg/sql/sql_activity_update_job.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfEy_R3TJqIVEx-sot:-NgfEy_R3TJqIVEx-sou:b-ryztnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L388)):*
> ```Go
>             merge_stats,
>             ''  AS query,
>             	(merge_stats -> 'statistics' ->> 'cnt')::int,
> ```

nit: seems like a few extra spaces upfront.
___
*[`pkg/sql/sql_activity_update_job.go` line 522 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfFlwpBvc7jZGkuGut:-NgfFlwpBvc7jZGkuGuu:b1glbj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L522)):*
> ```Go
>        merged_stats,
>        max_plan,
> 			 (select COALESCE(array_agg(o.rec::string), (array[]::string[])) FROM jsonb_array_elements_text(merged_stats -> 'index_recommendations') o(rec)) as idx_rec,
> ```

nit: indentation.
___
*[`pkg/sql/sql_activity_update_job.go` line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfFodr8lvQSWeJCHJw:-NgfFodr8lvQSWeJCHJx:b1glbj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L537)):*
> ```Go
>              max(ss.agg_interval) AS max_agg_interval,
>              max(ss.plan) AS max_plan,
> 						 merge_stats_metadata(ss.metadata) AS metadata,
> ```

nit: indentation.
___
*[`pkg/sql/sql_activity_update_job.go` line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfFqRBEpEYkNeiyq9G:-NgfFqRBEpEYkNeiyq9H:b-hbgen2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L540)):*
> ```Go
>     				 merge_statement_stats(ss.statistics) AS merged_stats
>       FROM system.statement_statistics ss
> 		  inner join limit_stmt_stats using (aggregated_ts, fingerprint_id, app_name)
> ```

nit: we usually capitalize INNER JOIN and GROUP BY
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGRUdAvll7qvsa_k0:-NgfGRUdAvll7qvsa_k1:bohme5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L347)):*
> ```
>                     │ group by: fingerprint_id, plan_hash, app_name
>                     │
>                     └── • apply join (left outer)
> ```

This apply join seems a bit concerning.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 491 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGZTK5zxEea2Qka_R:-NgfGZTK5zxEea2Qka_S:b1glbj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L491)):*
> ```
>                                  app_name,
>                                  fingerprint_id),
>      limit_stmt_stats AS (SELECT aggregated_ts,
> ```

nit: indentation.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-NgfGd5P-sq46AtHkK1i:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
> ```
>                     │ group by: fingerprint_id, plan_hash, app_name
>                     │
>                     └── • apply join (left outer)
> ```

Ditto.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 885 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGiyxCf_S5zn49qzI:-NgfGiyxCf_S5zn49qzJ:b1glbj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L885)):*
> ```
> ----
> distribution: local
>             vectorized: true
> ```

nit: indentation.
___
*[`pkg/sql/sql_activity_update_job_test.go` line 379 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGwgP1e86ONswn7Xy:-NgfGwgP1e86ONswn7Xz:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job_test.go#L379)):*
> ```Go
> 			    WHERE app_name = $3;`, 10000, 1, "topTransaction")
> 
> 		// Help check for primary key conflicts
> ```

nit: missing period.


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @yuzefovich)

___
*[`-- commits` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Ngf8th60zvxbJ91tQJ6):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This fixes #111224, right?
</blockquote></details>

Right, added it
___
*[`pkg/server/combined_statement_stats.go` line 1016 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Ngf8b0b9pbhUoCnUqLi:-NgfNAjZBOmGevLLr1L3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/server/combined_statement_stats.go#L1016)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: "transaction have".
</blockquote></details>

Done
___
*[`pkg/sql/sql_activity_update_job.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfEojI6wALYxKzhqDV:-NgfNDyk3jO72rDV7qNq:bv2mu0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L368)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why did this change? Because we only need 5 columns? Also, do we need to reduce `numberOfTopColumns` to `5` too? Also, `s/3000/2500/`.
</blockquote></details>

For Stmt we have 6 and for Txn we not have 5, so I updated the comment here to 2500 since is about Txn. I also created separate `numberOfTopColumns` for stmt and txn to make the distinction clear
___
*[`pkg/sql/sql_activity_update_job.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfEy_R3TJqIVEx-sot:-NgfNQTz31SCvFT1gnvh:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L388)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: seems like a few extra spaces upfront.
</blockquote></details>

Fixed
___
*[`pkg/sql/sql_activity_update_job.go` line 522 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfFlwpBvc7jZGkuGut:-NgfNTUWES0oQCHlIrHY:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L522)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: indentation.
</blockquote></details>

Fixed
___
*[`pkg/sql/sql_activity_update_job.go` line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfFodr8lvQSWeJCHJw:-NgfNV8S1A5x2QE4NCoP:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L537)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: indentation.
</blockquote></details>

Fixed
___
*[`pkg/sql/sql_activity_update_job.go` line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfFqRBEpEYkNeiyq9G:-NgfNXB32StjgKO3G8Y-:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job.go#L540)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we usually capitalize INNER JOIN and GROUP BY
</blockquote></details>

Fixed
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 347 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGRUdAvll7qvsa_k0:-NgfNa0P0Uu83BFRRntD:basey21) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L347)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This apply join seems a bit concerning.
</blockquote></details>

any suggestions?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 491 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGZTK5zxEea2Qka_R:-NgfNl2DAyOKOlKTQVLl:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L491)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: indentation.
</blockquote></details>

Fixed
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-NgfNmOkCo970DFfFApY:basey21) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

any suggestions?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 885 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGiyxCf_S5zn49qzI:-NgfNq_xEZGtyBW5hk01:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L885)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: indentation.
</blockquote></details>

Fixed
___
*[`pkg/sql/sql_activity_update_job_test.go` line 379 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGwgP1e86ONswn7Xy:-NgfNx-GCWjxQ8wWx1aj:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/sql_activity_update_job_test.go#L379)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: missing period.
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 2 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @maryliag)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-NgtZ0hL-H3X3VGn1JcP:b-p0dqhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

any suggestions?
</blockquote></details>

```sql
(select COALESCE(array_agg(o.rec::string), (array[]::string[])) FROM jsonb_array_elements_text(merged_stats -> 'index_recommendations') o(rec)) as idx_rec,
```
is what's causing us to use the apply join here. What we're trying to do is to array aggregate all index recommendations.

However, if we look at how `merge_statement_stats` does aggregation of index recommendations, `StatementStatistics.Add` simply keeps the latest `IndexRecommendations` as the code is written currently. Thus, I think we can change the query to simply use
```sql
merged_stats -> 'index_recommendations' as idx_rec,
```
which removes the need for apply join. WDYT?


<!-- Sent from Reviewable.io -->

maryliag committed:
sql: builtin for jsonb array to string array conversion

This commit introduces the function
`jsonb_array_to_string_array` that converts a JSONB array
to a string array format.

Epic: none

Release note (sql change): New builtin `jsonb_array_to_string_array`
that converts JSONB array to string array.

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-Ngym3TP9bq1l83h49bX:bgdq28i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

```sql
(select COALESCE(array_agg(o.rec::string), (array[]::string[])) FROM jsonb_array_elements_text(merged_stats -> 'index_recommendations') o(rec)) as idx_rec,
```
is what's causing us to use the apply join here. What we're trying to do is to array aggregate all index recommendations.

However, if we look at how `merge_statement_stats` does aggregation of index recommendations, `StatementStatistics.Add` simply keeps the latest `IndexRecommendations` as the code is written currently. Thus, I think we can change the query to simply use
```sql
merged_stats -> 'index_recommendations' as idx_rec,
```
which removes the need for apply join. WDYT?
</blockquote></details>

Query is now using the new builtin and the plan is updated. Let me know what you think


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 10 of 10 files at r4, 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @maryliag)

___
*[`-- commits` line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3QTX02nRMVlZ47-KU):*
nit: probably this sentence is no longer correct - we now use the last index recommendations that we happen to see when merging.
___
*[`pkg/sql/sql_activity_update_job.go` line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3QkPg0jBLMP1GMT25:-Nh3QkPg0jBLMP1GMT26:bjjqq6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/sql_activity_update_job.go#L68)):*
> ```Go
> 	settings.WithPublic)
> 
> const numberOfStmtTopColumns = 6
> ```

Should one of these be 5?
___
*[`pkg/sql/appstatspb/app_stats.go` line 200 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3RE145KXTtNqcfS6d:-Nh3RE145KXTtNqcfS6e:bk48tj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/appstatspb/app_stats.go#L200)):*
> ```Go
> 
> 	// Use the LastExecTimestamp to decide which object has the last error
> 	// and index recommendations.
> ```

nit: the comment needs an update.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-Nh3SvpqAyvXlcya3xVs:b76u79h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

Query is now using the new builtin and the plan is updated. Let me know what you think
</blockquote></details>

The plans no longer seem concerning.

Ideally, we'd run EXPLAIN ANALYZE of the read portions of these queries (similar to what I did [here](https://github.com/cockroachlabs/support/issues/2617#issuecomment-1731923133)) on a reasonably active database to see the performance. I guess it requires usage of multiple different fingerprints - perhaps we could use `sqlsmith` to generate such random queries.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 251 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3RaMh2b-4njBTFj6A:-Nh3RaMh2b-4njBTFj6B:bqgo5vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/opt/exec/execbuilder/testdata/observability#L251)):*
> ```
> query T retry
> EXPLAIN (VERBOSE)
>       UPSERT
> ```

nit: perhaps squash `UPSERT` with the next line?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 806 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3RsP29QM4xI4lCyZ7:-Nh3RsP29QM4xI4lCyZ8:b-22yz62) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/opt/exec/execbuilder/testdata/observability#L806)):*
> ```
>             merge_stats,
>             ''  AS query,
>               (merge_stats -> 'statistics' ->> 'cnt')::int,
> ```

nit: indentation is a bit off.


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @yuzefovich)

___
*[`-- commits` line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3QTX02nRMVlZ47-KU):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: probably this sentence is no longer correct - we now use the last index recommendations that we happen to see when merging.
</blockquote></details>

Fixed
___
*[`pkg/sql/sql_activity_update_job.go` line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3QkPg0jBLMP1GMT25:-Nh3bDlxEcAJ9PsrHwvi:bb9r7lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/sql_activity_update_job.go#L68)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Should one of these be 5?
</blockquote></details>

:facepalm: duh, fixed
___
*[`pkg/sql/appstatspb/app_stats.go` line 200 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3RE145KXTtNqcfS6d:-Nh3bTAXE3y7uFXqJQZh:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/appstatspb/app_stats.go#L200)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the comment needs an update.
</blockquote></details>

Done
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-Nh3c-FC7QM6FeoJZ4wF:b-q06qxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The plans no longer seem concerning.

Ideally, we'd run EXPLAIN ANALYZE of the read portions of these queries (similar to what I did [here](https://github.com/cockroachlabs/support/issues/2617#issuecomment-1731923133)) on a reasonably active database to see the performance. I guess it requires usage of multiple different fingerprints - perhaps we could use `sqlsmith` to generate such random queries.
</blockquote></details>

isn't this what the above part of `INJECT STATISTICS` is trying to simulate? We are adding stats to pretend those tables are full
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 251 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3RaMh2b-4njBTFj6A:-Nh3bfFAEIEhAALZUe3G:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/opt/exec/execbuilder/testdata/observability#L251)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps squash `UPSERT` with the next line?
</blockquote></details>

Done
___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 806 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3RsP29QM4xI4lCyZ7:-Nh3cCOh7mZbnsgEPTJQ:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/285bc78f57da25774c153653b5eb043f15bbb7df/pkg/sql/opt/exec/execbuilder/testdata/observability#L806)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: indentation is a bit off.
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 6 of 6 files at r6, 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @maryliag)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-Nh3gXZO5xgghNIVcI6P:b-n2trzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

isn't this what the above part of `INJECT STATISTICS` is trying to simulate? We are adding stats to pretend those tables are full
</blockquote></details>

Yes, this shows the right logical plan, but I'm wondering whether the actual performance (thus `EXPLAIN ANALYZE`) is fast enough. It's hard to say that simply by glancing at the plan. I mean, it looks like the plan has improved, but it would still be nice to see the actual runtime of the queries, especially so given that we'd be backporting this to 23.1.x I think.
___
*[`pkg/sql/sql_activity_update_job_test.go` line 99 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3gxKv1CgkeV8sUHnY:-Nh3gxKv1CgkeV8sUHnZ:b-z0ixob) ([raw file](https://github.com/cockroachdb/cockroach/blob/84ce54f739b600ab0e905f169b874ef69951f5f0/pkg/sql/sql_activity_update_job_test.go#L99)):*
> ```Go
> 	verifyActivityTablesAreEmpty(t, db)
> 
> 	// Us random name to keep isolated during stress tests.
> ```

nit: `s/Us/Use/`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Probably want to backport this to 23.2 and 23.1, right?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-:-Nh4VGKGEBmFosl0Kz6z:bjbwk79)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @maryliag)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/observability` line 606 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-NgfGd5P-sq46AtHkK1h:-Nh4VM2y8L7h1-j1Jj_U:b-ppjz7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3055a702a5a615c53dad86ba81f0d9d03f2d3953/pkg/sql/opt/exec/execbuilder/testdata/observability#L606)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yes, this shows the right logical plan, but I'm wondering whether the actual performance (thus `EXPLAIN ANALYZE`) is fast enough. It's hard to say that simply by glancing at the plan. I mean, it looks like the plan has improved, but it would still be nice to see the actual runtime of the queries, especially so given that we'd be backporting this to 23.1.x I think.
</blockquote></details>

Oh, I forgot that we already addressed - probably - the main reason for slowness we observed in a customer ticket (#111303), and this PR seems to improve the plan further, so I no longer think that we _need_ to run EXPLAIN ANALYZE to see the actual performance, so I'm ok with merging this.

That said, if you do see an easy / quick way to replicate the customer ticket that found these problems, getting a stmt bundle to see whether further improvements can be made would be a nice-to-have.


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
Right, I added the label for those backports

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-:-Nh8Rp-31mItrgzLdsIs:b-k4sqe4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian and @yuzefovich)

___
*[`pkg/sql/sql_activity_update_job_test.go` line 99 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/112350#-Nh3gxKv1CgkeV8sUHnY:-Nh8RnCV0rIub-Z_bQ4D:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/84ce54f739b600ab0e905f169b874ef69951f5f0/pkg/sql/sql_activity_update_job_test.go#L99)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/Us/Use/`.
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3546,6 +3546,37 @@ value if you rely on the HLC for accuracy.`,
+				}
+				jsonArray, ok := tree.MustBeDJSON(args[0]).AsArray()
+				if !ok {
+					return nil, pgerror.Newf(pgcode.InvalidParameterValue, "input argument must by JSON array type")
```
super nit: by -> be

maryliag replied:
Fixed on the original PR and rebased here

rafiss commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3546,6 +3546,37 @@ value if you rely on the HLC for accuracy.`,
 		},
 	),
 
+	"jsonb_array_to_string_array": makeBuiltin(arrayProps(),
```
out of curiosity, did you consider implementing this using a SQL body instead? for example:
```
root@localhost:26257/system/defaultdb> select array_agg(x) from (select jsonb_array_elements_text('[1,2,3]') as x);
  array_agg
-------------
  {1,2,3}
```

it might be faster in some cases, but more importantly, it might be easier to maintain.

maryliag replied:
That is how I had initially in this PR, but then Yahor [noticed](https://github.com/cockroachdb/cockroach/pull/112350#pullrequestreview-1680613676) it was causing an apply join that could be concerning, so this is why the suggestion to create the new builtin (which was reviewed on https://github.com/cockroachdb/cockroach/pull/112609 and rebased here so it could make sure it was working as expected)

rafiss replied:
oh got it, ignore me then! at least i learned something :)

maryliag replied:
No worries! I also learned 😄

maryliag commented:
TFTRs! I'll go ahead and merge this now

bors r+
